src/Workspaces/Core/Portable/Shared/Extensions/ParenthesizedExpressionSyntaxExtensions.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.Extensions
+{
+    internal static class CommonParenthesizedExpressionSyntaxExtensions
+    {
+        public static bool IsSafeToChangeAssociativity(
+            this SyntaxNode parenthesizedExpression, SyntaxNode innerExpression, SemanticModel semanticModel)
+        {
+            // Now we'll perform a few semantic checks to determine whether removal 
+            // of the parentheses might break semantics. Note that we'll try and be 
+            // fairly conservative with these. For example, we'll assume that failing 
+            // any of these checks results in the parentheses being declared as necessary 
+            // -- even if they could be removed depending on whether the parenthesized
+            // expression appears on the left or right side of the parent binary expression.
+
+            // First, does the binary expression result in an operator overload being 
+            // called?
+            var symbolInfo = semanticModel.GetSymbolInfo(innerExpression);
+            if (AnySymbolIsUserDefinedOperator(symbolInfo))
+            {
+                return false;
+            }
+
+            // Second, check the type and converted type of the binary expression.
+            // Are they the same?
+            var innerTypeInfo = semanticModel.GetTypeInfo(innerExpression);
+            if (innerTypeInfo.Type != null && innerTypeInfo.ConvertedType != null)
+            {
+                if (!innerTypeInfo.Type.Equals(innerTypeInfo.ConvertedType))
+                {
+                    return false;
+                }
+            }
+
+            // Floating point is not safe to change associativity of.  For example,
```

Comment:
:memo: (for other reviewers) This section is new

❓ Does this need to special-case the handling of `int + (int + long)`?

Reply:
Probably.  That may be caught by the above check.  But it may not be.  I'm going to add a check that all the types have to be the same.

Reply:
#done.